### DISCUSSION TITLE
Update subset of vertices of a Mesh

### DISCUSSION THREAD

**Discussion by AlgoryxJosef** [RobotecAI/RobotecGPULidar#242] [General]
If I have understood the API correctly, it is currently only possible to update ALL vertices of a mesh at once.
We have a situation where we have a very large mesh where a few vertices will change often.

I have a few questions:

1. Is it possible to handle this with the current API while getting good performance?
2. If not, do you have any hints on how the implementation may be changed to support updating a subset of vertices of a mesh? 

Regarding question 2 I assume the difficult thing is that vertex data is written to the GPU, is it hard to update a subset of vertices there?

Keep up the good work,
Best regards,
Josef

**Comment by prybicki** [comment]
Hi @AlgoryxJosef!

> If I have understood the API correctly, it is currently only possible to update ALL vertices of a mesh at once.

Yes, your understanding is correct. It was developed as an easier alternative to GPU-skinning, which we would like to implement in the future.

> Is it possible to handle this with the current API while getting good performance?

1. An easy thing to try would be to replace `OPTIX_BUILD_FLAG_PREFER_FAST_TRACE` with `OPTIX_BUILD_FLAG_PREFER_FAST_BUILD` [here](https://github.com/RobotecAI/RobotecGPULidar/blob/develop/src/scene/Mesh.cpp#L116) and recompiling the library.
2. You could also consider splitting the mesh on the client side to put the fast-changing parts into a separate mesh.

> If not, do you have any hints on how the implementation may be changed to support updating a subset of vertices of a mesh?

This could be achieved by adding another API call which would take only the subset of vertices (be it continguous range or a list of indices) and [copy](https://github.com/RobotecAI/RobotecGPULidar/blob/develop/src/scene/Mesh.cpp#L31) only those vertices. However, while this method would reduce the amount of data copied to the GPU, it would not change the cost of [re-building OptiX acceleration structure](https://github.com/RobotecAI/RobotecGPULidar/blob/develop/src/scene/Mesh.cpp#L54), which may be actually bigger than the copy itself. Optimizing the OptiX AS re-build is far more difficult.

> Keep up the good work

Thank you! We'd be happy to hear more details about your use-case, if possible.
Also, please feel free to reach out for commercial support to adam.dabrowski@robotec.ai.

**Reply by AlgoryxJosef** [reply]
Thank you for the quick answer, it is really helpful.
I don't have any further questions at the moment, but may have in the future.
All the best,
Josef
